pyccel/stdlib/ndarrays/ndarrays.h

```diff
@@ -32,6 +32,34 @@ typedef struct  s_slice
     int32_t step;
 }               t_slice;
 
+#define GET_INDEX_EXP1(t, arr,ndim, a) t(arr,ndim, a)
+#define GET_INDEX_EXP2(t, arr,ndim, a, b) GET_INDEX_EXP1(t, arr,ndim - 1, a) + t(arr,ndim, b)
+#define GET_INDEX_EXP3(t, arr,ndim, a, b, c) GET_INDEX_EXP2(t, arr,ndim - 1, a, b) + t(arr,ndim, c)
+#define GET_INDEX_EXP4(t, arr,ndim, a, b, c, d) GET_INDEX_EXP3(t, arr,ndim - 1, a, b, c) + t(arr,ndim, d)
+#define GET_INDEX_EXP5(t, arr,ndim, a, b, c, d, e) GET_INDEX_EXP4(t, arr,ndim - 1, a, b, c, d) + t(arr,ndim, e)
+#define GET_INDEX_EXP6(t, arr,ndim, a, b, c, d, e, f) GET_INDEX_EXP5(t, arr,ndim - 1, a, b, c, d, e) + t(arr,ndim, f)
+#define GET_INDEX_EXP7(t, arr,ndim, a, b, c, d, e, f, g) GET_INDEX_EXP6(t, arr,ndim - 1, a, b, c, d, e, f) + t(arr,ndim, g)
+#define GET_INDEX_EXP8(t, arr,ndim, a, b, c, d, e, f, g, h) GET_INDEX_EXP7(t, arr,ndim - 1, a, b, c, d, e, f, g) + t(arr,ndim, h)
+#define GET_INDEX_EXP9(t, arr,ndim, a, b, c, d, e, f, g, h, i) GET_INDEX_EXP8(t, arr,ndim - 1, a, b, c, d, e, f, g, h) + t(arr,ndim, i)
+#define GET_INDEX_EXP10(t, arr,ndim, a, b, c, d, e, f, g, h, i, j) GET_INDEX_EXP9(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i) + t(arr,ndim, j)
+#define GET_INDEX_EXP11(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k) GET_INDEX_EXP10(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j) + t(arr,ndim, k)
+#define GET_INDEX_EXP12(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k, l) GET_INDEX_EXP11(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j, k) + t(arr,ndim, l)
+#define GET_INDEX_EXP13(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k, l, m) GET_INDEX_EXP12(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j, k, l) + t(arr,ndim, m)
+#define GET_INDEX_EXP14(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k, l, m, n) GET_INDEX_EXP13(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j, k, l, m) + t(arr,ndim,n)
+#define GET_INDEX_EXP15(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) GET_INDEX_EXP14(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j, k, l, m, n) + t(arr,ndim,o)
+#define GET_INDEX_EXP16(t, arr,ndim, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p) GET_INDEX_EXP15(t, arr,ndim - 1, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) + t(arr,ndim,p)

```

Comment:
I wonder if this could be made clearer written in the opposite direction. Ie:
```
#define GET_INDEX_EXP1(t, arr,ndim, a) t(arr,ndim, a)
#define GET_INDEX_EXP2(t, arr,ndim, a, b) t(arr,ndim, a) + GET_INDEX_EXP1(t, arr,ndim - 1, b)
#define GET_INDEX_EXP3(t, arr,ndim, a, ...) t(arr,ndim, a) + GET_INDEX_EXP2(t, arr,ndim - 1, __VA_ARGS__)
...
#define GET_INDEX_EXP16(t, arr,ndim, a, ...) t(arr,ndim, a) + GET_INDEX_EXP15(t, arr,ndim - 1, __VA_ARGS__)
```

Reply:
I was comparing to the preprocessed code so it made sense to me to go from the index 0 to n.

Reply:
I think this should give the same code.
Imagine:
```
GET_INDEX_EXP3(t, arr,3, 4, 5, 6)
```
Before it gave us:
```
GET_INDEX_EXP2(t, arr, 2, 4, 5) + t(arr, ndim, 6) = GET_INDEX_EXP1(t, arr, 1, 4) + t(arr, ndim, 5) + t(arr, ndim, 6) = t(arr, ndim, 4) + t(arr, ndim, 5) + t(arr, ndim, 6)
```
With my modification it gives us:
```
t(arr, ndim, 4) + GET_INDEX_EXP2(t, arr, 2, 5, 6) = t(arr, ndim, 4) + t(arr, ndim, 5) + GET_INDEX_EXP1(t, arr, 1, 6) = t(arr, ndim, 4) + t(arr, ndim, 5) + t(arr, ndim, 6)
```

Reply:
Ah! No sorry you're right the `ndim` argument will be wrong

Reply:
But why do we need to pass this? Is it not known? Could we not do either:
```
#define GET_INDEX_EXP1(t, arr, a) t(arr,0, a)
#define GET_INDEX_EXP2(t, arr, a, b) GET_INDEX_EXP1(t, arr,a) + t(arr,1, b)
#define GET_INDEX_EXP3(t, arr, a, b, c) GET_INDEX_EXP2(t, arr,a, b) + t(arr,2, c)
...
#define GET_INDEX_EXP16(t, arr, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p) GET_INDEX_EXP15(t, arr, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) + t(arr,15,p)
```
or
```
#define GET_INDEX_EXP1(t, arr,ndim, a) t(arr,ndim-1, a)
#define GET_INDEX_EXP2(t, arr,ndim, a, b) t(arr,ndim-2, a) + GET_INDEX_EXP1(t, arr,ndim, b)
#define GET_INDEX_EXP3(t, arr,ndim, a, ...) t(arr,ndim-3, a) + GET_INDEX_EXP2(t, arr,ndim, __VA_ARGS__)
...
#define GET_INDEX_EXP16(t, arr,ndim, a, ...) t(arr,ndim-16, a) + GET_INDEX_EXP15(t, arr,ndim, __VA_ARGS__)
```
?

Reply:
I think removing the `ndim` better so it is readable even after the preprocessing

Reply:
Don't they both generate the same code?

Reply:
its the same but when having the `ndim` then preprocessed code will still have [ndim - 1 - 1 - 1] but we can have the exact number.

Reply:
Ah ok, then yes your solution is better. It also seems to be compiling faster now so the tests aren't slower